### PR TITLE
expression: implement vectorized evaluation for `builtinCastTimeAsStringSig`

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
Implement vectorized evaluation for builtinCastTimeAsStringSig, for #12106

### What is changed and how it works?
```
BenchmarkVectorizedBuiltinCastFunc/builtinCastTimeAsStringSig-VecBuiltinFunc-8         	    3000	    459010 ns/op	  129573 B/op	    7990 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastTimeAsStringSig-NonVecBuiltinFunc-8      	    3000	    478030 ns/op	  129572 B/op	    7990 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastTimeAsStringSig-VecBuiltinFunc#01-8      	    3000	    477695 ns/op	  129092 B/op	    8014 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastTimeAsStringSig-NonVecBuiltinFunc#01-8   	    3000	    475951 ns/op	  129092 B/op	    8014 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -577,11 +577,49 @@ func (b *builtinCastTimeAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk
...
-	return false
+	return true
 }
 
 func (b *builtinCastTimeAsStringSig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	var buf *chunk.Column
+	var err error
+	if b.args[0].GetType().Tp == mysql.TypeTimestamp {
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: Why do we check this?
The same buf will be returned for ETTimestamp and ETDatetime.

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -577,11 +577,49 @@ func (b *builtinCastTimeAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk
...
+
+	sc := b.ctx.GetSessionVars().StmtCtx
+	vas := buf.Times()
+	result.ReserveString(n)
+	for i, v := range vas {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		res, e := types.ProduceStrWithSpecifiedTp(v.String(), b.tp, sc, false)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: s/ e/ err

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -577,11 +577,49 @@ func (b *builtinCastTimeAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk
...
+		if e != nil {
+			return e
+		}
+		str, b, e1 := padZeroForBinaryType(res, b.tp, b.ctx)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: str, b, e1 ==> res, isNull, err

File: expression/builtin_cast_vec.go [hunk 4]
```diff
@@ -577,11 +577,49 @@ func (b *builtinCastTimeAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chunk
...
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalTime(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	sc := b.ctx.GetSessionVars().StmtCtx
+	vas := buf.Times()
+	result.ReserveString(n)
+	for i, v := range vas {
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: `v` is never used. We can use `for i:=0; i < n; i++`

File: expression/builtin_cast_vec.go (hunk 1 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, got it~

File: expression/builtin_cast_vec.go (hunk 2 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, i will replace it

Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, got it~

File: expression/builtin_cast_vec.go (hunk 3 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, got it~

File: expression/builtin_cast_vec.go (hunk 4 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: v is used. 
res, err = types.ProduceStrWithSpecifiedTp(v.String(), b.tp, sc, false)

**Pull request merged**